alf/algorithms/oac_algorithm.py

```diff
@@ -163,12 +159,40 @@ def mean_shift_fn(mu, dqda, sigma):
                 lambda spec: spec.sample(outer_dims=outer_dims),
                 self._action_spec)
         elif rollout and self._training_started:
-            critic_action = normal_dist.mean.detach().clone()
-            critic_action.requires_grad = True
-            transformed_action = critic_action
+            normal_dist = nest.map_structure(lambda dist: dist.base_dist,
+                                             action_dist)
+            unsquashed_mean = nest.map_structure(lambda dist: dist.mean,
+                                                 normal_dist)
+            unsquashed_std = nest.map_structure(lambda dist: dist.stddev,
+                                                normal_dist)
+            unsquashed_var = nest.map_structure(lambda dist: dist.variance,
+                                                normal_dist)
+
+            def _prepare_critic_action(dist_mean):
+                critic_action = dist_mean.detach().clone()
+                critic_action.requires_grad = True
+                return critic_action
+
+            def _prepare_tranform_action(action, dist):
+                for transform in dist.transforms:
+                    transformed_action = transform(action)
+                return transformed_action
+
+            def mean_shift_fn(mu, dqda, sigma):
+                if self._dqda_clipping:
+                    dqda = torch.clamp(dqda, -self._dqda_clipping,
+                                       self._dqda_clipping)
+                norm = torch.sqrt(torch.sum(torch.mul(dqda * dqda,
+                                                      sigma))) + 1e-6
+                shift = self._explore_delta * torch.mul(sigma, dqda) / norm
+                return mu + shift
+
+            critic_action = nest.map_structure(_prepare_critic_action,
```

Comment:
Is it possible to implement a ``MeanShiftDistributionTransformation``, which will include ``mean_shift_fn`` inside? In that way, the code can be simplified since we don't need to manually do the process of mean_shift -> distribution transform?

Reply:
Updated according to our discussion. Many thanks for the constructive feedback!